lib/Dialect/FIRRTL/Transforms/CheckCombLoops.cpp

```diff
@@ -25,533 +26,454 @@
+      blockArgToFields[arg].insert(ref.getFieldID());
+  }
+
+  void forallRefersTo(Value val, std::function<LogicalResult(Node refNode)> f,
+                      bool baseCase = true) {
+    if (baseCase && f(Node(val, 0)).failed())
+      return;
+    auto refersToIter = valRefersTo.find(val);
+    if (refersToIter != valRefersTo.end())
+      for (auto ref : refersToIter->second)
+        if (f(ref).failed())
+          return;
+  }
+
+  void dump() {
+    LLVM_DEBUG({
+      for (auto valRef : valRefersTo) {
+        llvm::dbgs() << "\n val :" << valRef.first;
+        for (auto node : valRef.second)
+          llvm::dbgs() << "\n Refers to :" << getFieldName(node).first;
+      }
+      for (auto dtv : fieldToVals) {
+        llvm::dbgs() << "\n Field :" << getFieldName(dtv.first).first
+                     << " ::" << dtv.first.getValue();
+        for (auto val : dtv.second)
+          llvm::dbgs() << "\n val :" << val;
+      }
+      for (auto p : portPaths) {
+        llvm::dbgs() << "\n Output port : " << getFieldName(p.first).first
+                     << " has comb path from :";
+        for (auto src : p.second)
+          llvm::dbgs() << "\n Input port : " << getFieldName(src).first;
+      }
+      for (auto b : blockArgToFields) {
+        llvm::dbgs() << "\n Block argument:"
+                     << getFieldName(Node(b.first, 0)).first;
+        for (auto fi : b.second)
+          llvm::dbgs() << "\n Block argument field :"
+                       << getFieldName(Node(b.first, fi)).first;
+      }
+    });
   }
 
-  FModuleOp module;
+  FModuleOp &module;
   InstanceGraph &instanceGraph;
-  // Set of visiting and visited nodes.
-  llvm::SmallDenseSet<Value> visiting, visited;
-  // Combinational paths between the ports of a module.
-  llvm::SmallDenseMap<Value, llvm::SmallDenseMap<size_t, SmallVector<Node>>>
-      &portPaths;
-  // Map of each unique storage location id and the list of SSA values that
-  // refer to it.
-  llvm::DenseMap<size_t, SmallVector<Value>> storageLocIdToValuesMap;
-  // Map of an SSA value to the list of storage location ids, that it can refer
-  // to.
-  llvm::DenseMap<Value, SmallVector<size_t>> valToStorageLocIdMap;
-  // For each leaf value, record the aggregate base value that it corresponds
-  // to.
-  llvm::DenseMap<Value, Value> leafToBaseValMap;
-  size_t maxStorageLocId = 0;
-  SmallVector<Value> currentPath;
-  bool printCycle = true;
-  bool detectedCycle = false;
+  // Map of a Value to all the FieldRefs that it refers to.
+  DenseMap<Value, SetOfNodes> valRefersTo;
+
+  DenseMap<Node, DenseSet<Value>> fieldToVals;
+  // Comb paths that exist between module ports. This is maintained across
+  // modules.
+  DenseMap<Node, SetOfNodes> &portPaths;
+  // Map of the fields of an aggregate BlockArgument/Ports that are used within
+  // a module. This is maintained across modules.
+  DenseMap<BlockArgument, std::set<unsigned>> &blockArgToFields;
```

Comment:
1) Why `std::set`-- especially for elements that are small, that can be rough (see https://llvm.org/docs/ProgrammersManual.html#set-like-containers-std-set-smallset-setvector-etc ).

2) This seems to only be used for `dump`-- maybe wrap it with `#ifndef NDEBUG` (perhaps refactoring to make this cleaner than having this in a bunch of places)?

Reply:
This was redundant, removed it now.